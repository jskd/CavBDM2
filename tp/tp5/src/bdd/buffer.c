/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
// Pour getline (standard GNU)
#define _GNU_SOURCE

#include "buffer.h"
#include "hexdump.h"
#include "quicksort.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * @brief Structure de donnée repésentant un buffer
 */
struct buf {
  /// Value: Valeur
  char* v;
  /// Size: Taille
  size_t s;
  /// Count: Nombre de valeur entrée
  size_t c;
};

/**
 * Creation d'un buffer
 * @param[in] size Taille du buffer
 */
struct buf* buf_create( size_t size ) {
  struct buf* buf= (struct buf*) malloc(sizeof(struct buf));
  buf->v= (char*) malloc(size);

  buf->s= size;
  buf->c=0;
  return buf;
}

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buf_destroy(struct buf* buf) {
  free(buf->v);
  free(buf);
}

/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[in] buf   buffer d'entrée
 * @param[in] value valeur à entrer
 * @return    0   succès
 *            -1  erreur: buffer plein
 */
char buf_put(struct buf* buf, char value) {
  if(buf->c >= buf->s)
    return -1;

  buf->v[buf->c]= value;
  buf->c++;

  return 0;
}

/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in] buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buf* storeFileBuffer(const char* file_name, size_t buffer_size) {
  FILE* file = fopen(file_name, "r");
  if (file == NULL)
    return NULL;

  struct buf* buf= buf_create( buffer_size );

  char* line = NULL;
  size_t len = 0;

  while((getline(&line, &len, file)) != -1) {
    buf_put(buf, line[0]);
  }

  fclose(file);

  return buf;
}

/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char writeBufferInFile(const char* file_name, const struct buf* buf) {

  FILE* file = fopen( file_name, "w");
  if (file == NULL)
    return -1;

  for(int i=0; i<buf->c; i++)
    fprintf(file, "%c\n", buf->v[i]);

  fclose(file);
  return 0;
}

/**
 * @brief Quick sort buffer
 * @param[in] buf buffer à trier
 */
void buf_quicksort(struct buf* buf) {
  quicksort(buf->v, 0, (buf->c) -1);
}

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buf_dump(const struct buf* buf) {
  printf ("buffer count: %zu\n", buf->c);
  printf ("buffer size: %zu\n", buf->s);
  hexDump("buffer value:", buf->v, buf->c);
}

/**
 * @brief Retourne la valeur à la position d'un buffer
 *
 * @param[in]  buf   Buffer
 * @param[in]  index Index of value
 *
 * @return Value
 *         Si -1: Erreur index > count
 */
char buf_val(const struct buf* buf, int index)
{
  if(index < buf->c)
    return buf->v[index];
  return -1;
}

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buf_count(const struct buf* buf)
{
  return buf->c;
}
