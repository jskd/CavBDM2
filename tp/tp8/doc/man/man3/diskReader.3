.TH "diskReader" 3 "Mardi 5 Décembre 2017" "CavBDM2 - BDD" \" -*- nroff -*-
.ad l
.nh
.SH NAME
diskReader
.SH SYNOPSIS
.br
.PP
.SS "Attributs publics"

.in +1c
.ti -1c
.RI "FILE ** \fBf\fP"
.br
.ti -1c
.RI "char ** \fBfp\fP"
.br
.ti -1c
.RI "size_t \fBf_count\fP"
.br
.in -1c
.SH "Documentation des données membres"
.PP 
.SS "FILE** diskReader::f"

.SS "size_t diskReader::f_count"

.SS "char** diskReader::fp"


.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour CavBDM2 - BDD à partir du code source\&.
