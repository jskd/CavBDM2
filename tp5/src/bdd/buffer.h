/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#ifndef SRC_BDD_BUFFER_H
#define SRC_BDD_BUFFER_H

#include <stdio.h>

/**
 * @brief Structure de donnée repésentant un buffer
 * Remarque: Structure caché! Ne pas initilialiser autrement
 * qu'avec les fonctions prévu à cet effet!
 */
struct buf;

/**
 * Creation d'un buffer
 * @param[in] size Taille du buffer
 */
struct buf* buf_create( size_t size );

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buf_destroy(struct buf* buf);

/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[in] buf   buffer d'entrée
 * @param[in] value valeur à entrer
 * @return    0   succès
 *            -1  erreur: buffer plein
 */
char buf_put(struct buf* buf, char value);

/**
 * @brief Quick sort buffer
 *
 * @param buf Trie le buffer en entrée
 */
void buf_quicksort(struct buf* buf);

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buf_dump(const struct buf* buf);

/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in] buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buf* storeFileBuffer(const char* file_name, size_t buffer_size);


/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char writeBufferInFile(const char* file_name, const struct buf*);

/**
 * @brief Retourne la valeur à la position d'un buffer
 *
 * @param[in]  buf   Buffer
 * @param[in]  index Index of value
 *
 * @return Value
 */
char buf_val(const struct buf* buf, int index);

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buf_count(const struct buf* buf);

#endif
