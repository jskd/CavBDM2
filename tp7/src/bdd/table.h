/**
 * TP n°: 6
 *
 * Titre du TP :
 *
 * Date : 21/10/17
 *
 * Nom : Lefranc
 * Prenom : Joaquim
 * email : lefrancjoaquim@gmail.com
 *
 * Nom : Skoda
 * Prenom : Jérôme
 * email : contact@jeromeskoda.fr
 *
 * Remarques :
 */
#ifndef SRC_BDD_TABLE_H
#define SRC_BDD_TABLE_H

#include <stdio.h>
#include "bufferExtended.h"

struct table;

struct table* table_create(size_t bucket, const char* directory);

void table_insertbuffer(struct table* tab, const struct buffer* buf);

void table_putBuffer(struct table* tab, const struct buffer* buf);

void table_destroy(struct table* tab) ;
size_t table_get_n_bucket(const struct table* tab);

int table_get_write_stat(const struct table* tab);
void table_fprint_stat(FILE* stream, const struct table* tab);

struct diskReader* create_disk_from_bucket(const struct table* tab, int indexBucket) ;
#endif
