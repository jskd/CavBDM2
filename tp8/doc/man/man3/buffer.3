.TH "buffer" 3 "Vendredi 1 Décembre 2017" "CavBDM2 - BDD" \" -*- nroff -*-
.ad l
.nh
.SH NAME
buffer \- Structure de donnée repésentant un buffer\&.  

.SH SYNOPSIS
.br
.PP
.SS "Attributs publics"

.in +1c
.ti -1c
.RI "char * \fBv\fP"
.br
.RI "Valeur\&. "
.ti -1c
.RI "size_t \fBs\fP"
.br
.RI "Taille\&. "
.ti -1c
.RI "size_t \fBc\fP"
.br
.RI "Nombre de valeur entrée\&. "
.ti -1c
.RI "size_t \fBdata_size\fP"
.br
.RI "Taille des données\&. "
.ti -1c
.RI "\fBbuffer_file_mode\fP \fBmode\fP"
.br
.ti -1c
.RI "int \fBread_counter\fP"
.br
.RI "compteur de Lecture (en octet) "
.ti -1c
.RI "int \fBrfile_counter\fP"
.br
.ti -1c
.RI "int \fBwrite_counter\fP"
.br
.RI "compteur d'écriture (en octet) "
.ti -1c
.RI "int \fBwfile_counter\fP"
.br
.in -1c
.SH "Description détaillée"
.PP 
Structure de donnée repésentant un buffer\&. 
.SH "Documentation des données membres"
.PP 
.SS "size_t buffer::c"

.PP
Nombre de valeur entrée\&. 
.SS "size_t buffer::data_size"

.PP
Taille des données\&. 
.SS "\fBbuffer_file_mode\fP buffer::mode"

.SS "int buffer::read_counter"

.PP
compteur de Lecture (en octet) 
.SS "int buffer::rfile_counter"

.SS "size_t buffer::s"

.PP
Taille\&. 
.SS "char* buffer::v"

.PP
Valeur\&. 
.SS "int buffer::wfile_counter"

.SS "int buffer::write_counter"

.PP
compteur d'écriture (en octet) 

.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour CavBDM2 - BDD à partir du code source\&.
