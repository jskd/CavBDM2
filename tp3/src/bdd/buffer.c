/**
* TP n°: 2
*
* Titre du TP : Merge Join
*
* Date : 13/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#include "buffer.h"
#include "hexdump.h"
#include "quicksort.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * @brief Structure de donnée repésentant un buffer
 */
struct buf {
  /// Value: Valeur
  char* v;
  /// Size: Taille
  size_t s;
  /// Count: Nombre de valeur entrée
  size_t c;
};

/**
 * Creation d'un buffer
 * @param[in] size Taille du buffer
 */
struct buf* buf_create( size_t size ) {
  struct buf* buf= (struct buf*) malloc(sizeof(struct buf));
  buf->v= (char*) malloc(size);

  buf->s= size;
  buf->c=0;
  return buf;
}

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buf_destroy(struct buf* buf) {
  free(buf->v);
  free(buf);
}

/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[in] buf   buffer d'entrée
 * @param[in] value valeur à entrer
 * @return    0   succès
 *            -1  erreur: buffer plein
 */
char buf_put(struct buf* buf, char value) {
  if(buf->c >= buf->s)
    return -1;

  buf->v[buf->c]= value;
  buf->c++;

  return 0;
}

/**
 * @brief Enregistre un le 1er caractére de chaque
 *        ligne d'un fichier dans un buffer
 *
 * @param[in]  Fichier d'entrée
 * @param[out] Buffer de sortie
 */
void storeFileBuffer(FILE* fp, struct buf* buf) {
  char* line = NULL;
  size_t len = 0;

  while((getline(&line, &len, fp)) != -1) {
    buf_put(buf, line[0]);
  }
}

/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in] buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buf* storeFileBufferOC(const char* file_name, size_t buffer_size) {
  FILE* file = fopen(file_name, "r");
  if (file == NULL)
    return NULL;

  struct buf* buf= buf_create( buffer_size );
  storeFileBuffer(file, buf);
  fclose(file);

  return buf;
}

/**
 * Ecrit un buffer dans un fichier
 *
 * @param[out] fp  fichier de sortie
 * @param[in]  buf fichier d'entrée
 */
void writeBufferInFile(FILE* fp, const struct buf* buf) {
  for(int i=0; i<buf->c; i++)
    fprintf(fp, "%c\n", buf->v[i]);
}

/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char writeBufferInFileOC(const char* file_name, const struct buf* buf) {

  FILE* file = fopen( file_name, "w");
  if (file == NULL)
    return -1;

  writeBufferInFile(file, buf);
  fclose(file);
  return 0;
}

/**
 * @brief Quick sort buffer
 * @param[in] buf buffer à trier
 */
void buf_quicksort(struct buf* buf) {
  quicksort(buf->v, 0, (buf->c) -1);
}

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buf_dump(const struct buf* buf) {
  printf ("buffer count: %zu\n", buf->c);
  printf ("buffer size: %zu\n", buf->s);
  hexDump("buffer value:", buf->v, buf->c);
}

/**
 * @brief Retourne la valeur à la position d'un buffer
 *
 * @param[in]  buf   Buffer
 * @param[in]  index Index of value
 *
 * @return Value
 *         Si -1: Erreur index > count
 */
char buf_val(const struct buf* buf, int index)
{
  if(index < buf->c)
    return buf->v[index];
  return -1;
}

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buf_count(const struct buf* buf)
{
  return buf->c;
}
