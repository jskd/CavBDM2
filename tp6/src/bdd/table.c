/**
 * TP n°: 6
 *
 * Titre du TP :
 *
 * Date : 21/10/17
 *
 * Nom : Lefranc
 * Prenom : Joaquim
 * email : lefrancjoaquim@gmail.com
 *
 * Nom : Skoda
 * Prenom : Jérôme
 * email : contact@jeromeskoda.fr
 *
 * Remarques :
 */
#include "table.h"
#include "buffer.h"
#include "bucket.h"
#include <limits.h>
#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#define BUCKET_FILE_EXTENSION ".txt"
/**
 * @brief Structure de donnée repésentant un buffer
 */
struct table {
  /// bucket
  struct bucket** b;
  /// n bucket
  size_t n_bucket;
};

static int _table_hash(const struct table* tab, short key) {
  return key % tab->n_bucket;
}

struct table* table_create(size_t n_bucket, const char* directory) {
  struct table* tab= (struct table*) malloc(sizeof(struct table));
  tab->b= (struct bucket**) malloc(n_bucket * sizeof(struct bucket*) );
  mkdir(directory, 0777);
  for(int indexBucket=0; indexBucket<n_bucket; indexBucket++) {
    char filename[PATH_MAX];
    sprintf(filename, "%s/%d.bucket.txt", directory, indexBucket);
    tab->b[indexBucket]= bucket_create(filename);
  }
  tab->n_bucket= n_bucket;
  return tab;
}

void table_putBuffer(struct table* tab, const struct buffer* buf) {
  if(buffer_mode(buf) == BUFFER_DECIMALS
    && buffer_datasize(buf) == sizeof(short))
  {
    for(int index_buf=0; index_buf< buffer_count(buf); index_buf++) {
      short key=0;
      char str[100];
      sprintf(str, "%d", key);
      bucket_puts( tab->b[ _table_hash(tab, key) ], str);
    }
  }
  else
  {
    printf("Erreur, table put non implementé pour ce type de buffer\n");
  }
}

void table_storeBucketInBuffer(struct table* tab, int index, struct buffer* buf)
{
  if(index > 0 && index < tab->n_bucket)
    buffer_read_file_from_descriptor(  bucket_getFile(tab->b[index]), buf);
}
