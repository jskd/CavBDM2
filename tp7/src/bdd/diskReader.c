/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/

// Pour DT_REG (standard GNU)
#define _GNU_SOURCE

#include "diskReader.h"
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>


struct diskReader {
  FILE** v;
  size_t s;
  size_t c;
};

static int _count_regular_file(DIR * dirp) {
  rewinddir(dirp); // si utilisé deux fois de suite
  int file_count = 0;
  struct dirent * entry;
  while ((entry = readdir(dirp)) != NULL) {
    if (entry->d_type == DT_REG) {
      file_count++;
    }
  }
  return file_count;
}

static char _store_file_descriptor(struct diskReader* disk, const char * dir, DIR * dirp) {
  struct dirent * entry;
  FILE** ptr= disk->v;
  char filename[PATH_MAX];
  rewinddir(dirp); // si utilisé deux fois de suite
  while ((entry = readdir(dirp)) != NULL) {
    if (entry->d_type == DT_REG) {
      if(disk->c == disk->s) // Overflow
        return -1;
      sprintf(filename, "%s/%s", dir, entry->d_name);
      *ptr= fopen(filename, "r");
      if(*ptr == NULL) // Open error
        return -2;
      ptr++;
      disk->c++;
    }
  }
  return 0;
}

struct diskReader* disk_r_create( const char * dir) {

  DIR* dirp = opendir(dir);
  if(dirp == NULL)
    return NULL;

  struct diskReader* disk= (struct diskReader*) malloc(sizeof(struct diskReader));

  disk->s=  _count_regular_file(dirp);
  disk->v= (FILE**) malloc( sizeof(FILE*) * disk->s );
  disk->c= 0;

  char error= _store_file_descriptor(disk, dir, dirp);
  closedir(dirp);

  if(error) {
    disk_r_destroy(disk);
    return NULL;
  }

  return disk;
}

size_t disk_r_count(const struct diskReader* disk) {
  return disk->c;
}

FILE* disk_r_item(const struct diskReader* disk, int index) {
  return disk->v[index];
}

void disk_r_destroy(struct diskReader* disk) {
  for(int i=0; i< disk_r_count(disk); i++)
    fclose(disk_r_item(disk, i));
  free(disk->v);
  free(disk);
}

void disk_r_storeContentInTable(const struct diskReader* disk, struct buffer* buf, struct table* tab) {
  for(int index_disk=0; index_disk<disk_r_count(disk); index_disk++) {
    buffer_read_file_from_descriptor( disk_r_item(disk, index_disk),  buf);
    table_putBuffer(tab, buf);
  }
}
