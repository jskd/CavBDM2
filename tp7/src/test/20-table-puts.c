#include "../bdd/buffer.h"
#include "../bdd/table.h"
#include "../bdd/diskReader.h"

// Buffer config
static const size_t _buf_size= 10;
static const size_t _data_lenght= sizeof(short);
static const int _buffer_type= BUFFER_DECIMALS;

// input disk config
static const char* _file_r= "res/test/20/R";

// table config
static const char* _file_table_r= "res/test/20/tableR";
static const size_t _n_bucket= 2;

int main(int argc, char** argv){

  struct diskReader* disk_r= disk_r_create(_file_r);
  if(disk_r == NULL) {
    printf("Erreur lors de la lecture de %s.\n", _file_r);
    return -1;
  }

  struct table* tab_r= table_create(_n_bucket, _file_table_r);
  if(tab_r == NULL) {
    printf("Erreur lors de la creation de %s.\n", _file_table_r);
    return -1;
  }

  struct buffer* buf_r  = buffer_create(_buf_size, _data_lenght, _buffer_type);

  disk_r_storeContentInTable(disk_r, buf_r, tab_r);

  for(int index_bucket=0; index_bucket < table_get_n_bucket(tab_r);  index_bucket++) {
    struct diskReader* disk_tab= create_disk_from_bucket(tab_r, index_bucket);
    disk_r_storeContentInTable(disk_tab, buf_r, tab_r);
    printf("Value bucket %d\n", index_bucket);
    buffer_printValue(buf_r);
    disk_r_destroy(disk_tab);
  }


  disk_r_destroy(disk_r);
  table_destroy(tab_r);
  buffer_destroy(buf_r);
}
