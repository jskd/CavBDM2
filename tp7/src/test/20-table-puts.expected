Value bucket 0
32490
-4370
13610
1030
-23830
Value bucket 1
-19181
10081
-6721
-18301
30461
